mono/metadata/object.c

```diff
@@ -4471,27 +4489,32 @@ mono_string_new_utf16 (MonoDomain *domain, const guint16 *text, gint32 len)
 MonoString *
 mono_string_new_size (MonoDomain *domain, gint32 len)
 {
-	MonoString *s;
-	MonoVTable *vtable;
-	size_t size = (sizeof (MonoString) + ((len + 1) * 2));
+	if (len == 0) {
+		g_assert (domain->empty_string);
+		return domain->empty_string;
+	} else {
+		MonoString *s;
+		MonoVTable *vtable;
+		size_t size = (sizeof (MonoString) + ((len + 1) * 2));
 
-	/* overflow ? can't fit it, can't allocate it! */
-	if (len > size)
-		mono_gc_out_of_memory (-1);
+		/* overflow ? can't fit it, can't allocate it! */
```

Comment:
Maybe, maybe not.  This isn't the change set that will go upstream so I'm not particularly worried about it.